File: python/reader.py
import re
from mal_types import Symbol

class Blank(Exception): pass

class Reader():
    def __init__(self, tokens, position=0):
        self.tokens = tokens
        self.position = position

    def next(self):
        self.position += 1
        return self.tokens[self.position-1]

    def peek(self):
        if len(self.tokens) > self.position:
            return self.tokens[self.position]
        else:
            return None

def tokenize(str):
    tre = re.compile(r"""[\s,]*(~@|[\[\]{}()'`~^@]|"(?:[\\].|[^\\"])*"?|;.*|[^\s\[\]{}()'"`@,;]+)""");
    return [t for t in re.findall(tre, str) if t[0] != ';']

def unescape(str):
    return str.replace('\\"', '"').replace('\\n', '\n').replace('\\\\', '\\')

def read_atom(reader):
    int_re = re.compile(r"-?[0-9]+$")
    token = reader.next()
    if re.match(int_re, token):     return int(token)
    elif token[0] == '"':           return unescape(token[1:-1])
    elif token == "nil":            return None
    elif token == "true":           return True
    elif token == "false":          return False
    else:                           return Symbol(token)

def read_list(reader):
    ast = []
    token = reader.next()
    token = reader.peek()
    while token != ")":
        if not token: raise Exception("expected ')', got EOF")
        ast.append(read_form(reader))
        token = reader.peek()
    reader.next()
    return ast

def read_form(reader):
    token = reader.peek()
    # list
    if token == ')':   raise Exception("unexpected ')'")
    elif token == '(': return read_list(reader)

    else:              return read_atom(reader);

def read_str(str):
    tokens = tokenize(str)
    if len(tokens) == 0: return None
    return read_form(Reader(tokens))

File: js/reader.js
var types = require('./types')

function Reader(tokens) {
    this.tokens = tokens
    this.position = 0
}
Reader.prototype.next = function () {
    return this.tokens[this.position++]
}
Reader.prototype.peek = function () {
    return this.tokens[this.position]
}

function tokenize(str) {
    var re = /[\s,]*(~@|[\[\]{}()'`~^@]|"(?:\\.|[^\\"])*"|;.*|[^\s\[\]{}('"`,;)]*)/g
    var results = []
    while ((match = re.exec(str)[1]) != '') {
        if (match[0] === ';') { continue }
        results.push(match)
    }
    return results
}

function unescape(str) {
    return str.replace(/\\"/g, '"')
              .replace(/\\n/g, "\n")
              .replace(/\\\\/g, "\\")
}

function read_atom (reader) {
    var token = reader.next()
    if (token.match(/^-?[0-9]+$/)) {
        return parseInt(token)
    } else if (token.match(/^"/)) {
        return unescape(token.slice(1,token.length-1))
    } else if (token === 'nil') {
        return null
    } else if (token === 'true') {
        return true
    } else if (token === 'false') {
        return false
    } else {
        return new types.Symbol(token)
    }
}

function read_list(reader) {
    var list = []
    var token = reader.next()
    while ((token = reader.peek()) !== ')') {
        if (!token) {
            throw new Error("expected ')', got EOF")
        }
        list.push(read_form(reader))
    }
    reader.next()
    return list
}

function read_form(reader) {
    var token = reader.peek()
    switch (token) {
    case ')': throw new Error("unexpected ')'")
    case '(': return read_list(reader)
    default:  return read_atom(reader)
    }
}

function read_str(str) {
    var tokens = tokenize(str)
    if (tokens.length === 0) { return null }
    return read_form(new Reader(tokens))
}

exports.read_str = read_str
